### PR TITLE
Manual bit manipulation

### PR DESCRIPTION
**Pull request by lior-dikstein**
## Pull Request Description:
Implemented the BitWidthConfig API to enable manual adjustment of bit-width settings for specific model layers. Currently, only manual activation bit-width selection is supported. This functionality is implemented through a new configuration class under CoreConfig, named BitWidthConfig.

To change the activation bit width, a filter is applied along with the desired bit width. For example, to set the activation bit width of all the convolution layers in the model to 4 bits, and a layer named "add_1" in the model to 16 bit:

```
from model_compression_toolkit.core import CoreConfig
from model_compression_toolkit.core.common.network_editors import NodeTypeFilter, NodeNameFilter
import torch

core_config = CoreConfig()
core_config.bit_width_config.set_manual_activation_bit_width(NodeTypeFilter(torch.nn.Conv2d), 4)
core_config.bit_width_config.set_manual_activation_bit_width(NodeNameFilter('add_1'), 16)

```

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/common/quantization/core_config.py [hunk 1]
```diff
@@ -22,20 +23,23 @@ class CoreConfig:
...
         Args:
             quantization_config (QuantizationConfig): Config for quantization.
             mixed_precision_config (MixedPrecisionQuantizationConfig): Config for mixed precision quantization.
             If None, a default MixedPrecisionQuantizationConfig is used.
+            bit_width_config (BitWidthConfig): Config for bit-width selection.
             debug_config (DebugConfig): Config for debugging and editing the network quantization process.
         """
-        self.quantization_config = quantization_config
-        self.debug_config = debug_config
+        self.quantization_config = QuantizationConfig() if quantization_config is None else quantization_config
```
Comment (Idan-BenAmi) [inline model_compression_toolkit/core/common/quantization/core_config.py:40]: why not using the standard default assignment?

**Review by Idan-BenAmi — APPROVED** [review]

**Comment by ofirgo** [comment]
> ## Pull Request Description:
> Implemented BitWidthConfig API to allow manual adjustment of bit width settings for specific model layers.

1. Specify that currently only activation bitwidth setting is supported
2. Add to the description details regarding the implementation - design choices,  modules that were need to change, etc.

File: model_compression_toolkit/core/common/quantization/core_config.py [hunk 2]
```diff
@@ -22,20 +23,23 @@ class CoreConfig:
...
+                 bit_width_config: BitWidthConfig = None,
+                 debug_config: DebugConfig = None
                  ):
         """
 
         Args:
             quantization_config (QuantizationConfig): Config for quantization.
             mixed_precision_config (MixedPrecisionQuantizationConfig): Config for mixed precision quantization.
             If None, a default MixedPrecisionQuantizationConfig is used.
+            bit_width_config (BitWidthConfig): Config for bit-width selection.
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/quantization/core_config.py:37]: "for **manual** bit-width selection"
otherwise, it sounds like this is the proper way to set a bit-width, which is not

File: model_compression_toolkit/core/common/quantization/set_node_quantization_config.py [hunk 3]
```diff
@@ -194,26 +204,47 @@ def _create_node_candidates_qc(qc: QuantizationConfig,
...
 
     Returns:
-        List of candidates of weights quantization configurations to set for a node.
+        List[CandidateNodeQuantizationConfig]: List of candidates of weights quantization configurations to set for a node.
     """
 
     candidates = []
     node_attrs_list = node.get_node_weights_attributes()
 
+    if node in nodes_to_manipulate_bit_widths.keys():
```
Comment (ofirgo) [inline model_compression_toolkit/core/common/quantization/set_node_quantization_config.py:229]: Please extract this part to a separate method "manually_overide_node_bitwidth" or something like that) and add lots of documentation explaining the purpose and the functionality

File: tests/pytorch_tests/model_tests/feature_models/manual_bit_selection.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        x = self.bn1(out1)
+        x = self.conv2(x)
+        x = self.relu(x)
+        x = x + out1
+        # Flatten the tensor
+        x = x.view(x.size(0), -1)
+        output = self.fc(x)
+        return output
+
+def get_layer_type_from_activation_quantizer(model, layer_name):
```
Comment (elad-c) [inline tests/pytorch_tests/model_tests/feature_models/manual_bit_selection.py:66]: Seems general enough to move this to tests/pytorch_tests/utils.py

File: tests/pytorch_tests/model_tests/feature_models/manual_bit_selection.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    layer_for_act_quant_name = layer_name.split('_' + ACTIVATION_HOLDER_QUANTIZER)[0]
+    for name, layer in model.named_modules():
+        if name == layer_for_act_quant_name:
+            if isinstance(layer, PytorchQuantizationWrapper):
+                return layer.layer
+            else:
+                return layer
+
+class BaseManualBitWidthSelectionTest(MixedPrecisionActivationBaseTest):
+    def __init__(self, unit_test):
```
Comment (elad-c) [inline tests/pytorch_tests/model_tests/feature_models/manual_bit_selection.py:76]: Is this needed when it only calls super init?

File: tests/pytorch_tests/model_tests/feature_models/manual_bit_selection.py
Comment (elad-c) [inline tests/pytorch_tests/model_tests/feature_models/manual_bit_selection.py:1]: Add comments to every test stating the purpose of the test.

File: model_compression_toolkit/core/common/quantization/bit_width_config.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        if len(bit_widths) > 1 and len(bit_widths) != len(filters):
+            Logger.critical(f"Configuration Error: The number of provided bit_width values {len(bit_widths)} "
+                            f"must match the number of filters {len(filters)}, or a single bit_width value "
+                            f"should be provided for all filters.")
+        elif len(bit_widths) == 1 and len(filters) > 1:
+            bit_widths = [bit_widths[0] for f in filters]
+        for bit_width, filter in zip (bit_widths, filters):
+            self.manual_bit_width_selection_list += [ManualBitWidthSelection(filter, bit_width)]
+
+    def get_nodes_to_manipulate_bit_widths(self, graph: Graph):
```
Comment (elad-c) [inline model_compression_toolkit/core/common/quantization/bit_width_config.py:74]: return type

**Review by elad-c — COMMENTED** [review]
Need more comments in tests.

File: model_compression_toolkit/core/common/quantization/set_node_quantization_config.py [hunk 7]
```diff
@@ -262,3 +249,51 @@ def _create_node_candidates_qc(qc: QuantizationConfig,
                                                            node_attrs_list))
 
     return candidates
+
+
+def filter_qc_options_with_manual_bit_width(
+        node: BaseNode,
+        node_qc_options_list: List[OpQuantizationConfig],
+        base_config: OpQuantizationConfig,
+        manual_bit_width_override: Optional[int],
```
Comment (elad-c) [inline model_compression_toolkit/core/common/quantization/set_node_quantization_config.py:258]: why Optional?

File: tests/pytorch_tests/model_tests/feature_models/manual_bit_selection.py [hunk 8]
```diff
@@ -147,29 +162,70 @@ def __init__(self, unit_test, filters, bit_widths):
...
                                 break
                         self.unit_test.assertTrue(layer.activation_holder_quantizer.num_bits == bit_width)
                     else:
+                        # make sure that the bit width of other layers was not changed.
                         self.unit_test.assertFalse(layer.activation_holder_quantizer.num_bits in bit_widths, msg=f"name {name}, layer.activation_holder_quantizer.num_bits {layer.activation_holder_quantizer.num_bits }, {self.bit_widths}")
 
 
 class Manual16BitTest(ManualBitWidthByLayerNameTest):
 
+    # def get_tpc(self):
```
Comment (elad-c) [inline tests/pytorch_tests/model_tests/feature_models/manual_bit_selection.py:190]: remove

**Review by elad-c — APPROVED** [review]

File: model_compression_toolkit/core/common/quantization/set_node_quantization_config.py (hunk 7 above)
Comment (lior-dikstein) [reply model_compression_toolkit/core/common/quantization/set_node_quantization_config.py:258]: It can be None
